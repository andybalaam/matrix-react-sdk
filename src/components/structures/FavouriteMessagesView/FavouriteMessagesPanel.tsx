/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient, MatrixEvent } from "matrix-js-sdk/src/matrix";
import React, { FC, useRef } from "react";

import { _t } from "../../../languageHandler";
import { FavouriteMessagesStore } from "../../../stores/FavouriteMessagesStore";
import ResizeNotifier from "../../../utils/ResizeNotifier";
import ScrollPanel from "../ScrollPanel";
import FavouriteMessagesHeader from "./FavouriteMessagesHeader";
import FavouriteMessagesTilesList from "./FavouriteMessagesTilesList";

interface IProps {
    favouriteMessageEvents: MatrixEvent[] | null;
    resizeNotifier?: ResizeNotifier;
    searchQuery: string;
    handleSearchQuery: (query: string) => void;
    cli: MatrixClient;

    // Provide this to use a different store for favourite messages
    // e.g. in tests. If not supplied, we use the global default.
    favouriteMessagesStore?: FavouriteMessagesStore;
}

const FavouriteMessagesPanel: FC<IProps> = (props: IProps) => {
    const favouriteMessagesPanelRef = useRef<ScrollPanel>();

    if (props.favouriteMessageEvents?.length === 0) {
        return (
            <>
                <FavouriteMessagesHeader
                    query={props.searchQuery}
                    handleSearchQuery={props.handleSearchQuery}
                    favouriteMessagesStore={props.favouriteMessagesStore}
                />
                <h2 className="mx_FavouriteMessages_emptyMarker">{_t("No Favourite Messages")}</h2>
            </>
        );
    } else {
        return (
            <>
                <FavouriteMessagesHeader
                    query={props.searchQuery}
                    handleSearchQuery={props.handleSearchQuery}
                    favouriteMessagesStore={props.favouriteMessagesStore}
                />
                <ScrollPanel
                    ref={favouriteMessagesPanelRef ?? null}
                    className="mx_RoomView_searchResultsPanel mx_FavouriteMessages_scrollPanel"
                    resizeNotifier={props.resizeNotifier}
                >
                    <FavouriteMessagesTilesList
                        cli={props.cli}
                        favouriteMessageEvents={props.favouriteMessageEvents}
                        favouriteMessagesPanelRef={favouriteMessagesPanelRef}
                        searchQuery={props.searchQuery}
                        favouriteMessagesStore={props.favouriteMessagesStore}
                    />
                </ScrollPanel>
            </>
        );
    }
};
export default FavouriteMessagesPanel;
