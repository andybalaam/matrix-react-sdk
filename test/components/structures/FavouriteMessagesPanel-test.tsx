/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { mocked, MockedObject } from "jest-mock";
import { EventType, MatrixClient, MatrixEvent, MsgType } from "matrix-js-sdk/src/matrix";
import { render } from "@testing-library/react";

import { stubClient } from "../../test-utils";
import { MatrixClientPeg } from "../../../src/MatrixClientPeg";
import FavouriteMessagesPanel from "../../../src/components/structures/FavouriteMessagesView/FavouriteMessagesPanel";
import SettingsStore from "../../../src/settings/SettingsStore";

describe("FavouriteMessagesPanel", () => {
    let cli: MockedObject<MatrixClient>;
    // let room: Room;
    const userId = "@alice:server.org";
    const roomId = "!room:server.org";
    const alicesFavouriteMessageEvent = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: userId,
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: "i am alice",
        },
        event_id: "$alices_message",
        origin_server_ts: 123214,
    });

    const bobsFavouriteMessageEvent = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: "@bob:server.org",
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: "i am bob",
        },
        event_id: "$bobs_message",
        origin_server_ts: 123213,
    });

    beforeEach(async () => {
        stubClient();
        cli = mocked(MatrixClientPeg.get());
        jest.spyOn(SettingsStore, "getValue").mockImplementation((setting) => setting === "feature_favourite_messages");
    });

    afterEach(async () => {
        jest.restoreAllMocks();
    });

    it("renders <FavouriteMessagesPanel /> component with empty or default props correctly", () => {
        const favouriteMessageEvents: MatrixEvent[] | null = null;
        const props = {
            favouriteMessageEvents,
            handleSearchQuery: jest.fn(),
            searchQuery: "",
            cli,
        };
        const panel = render(<FavouriteMessagesPanel {...props} />);
        expect(panel.findByText("No Favourite Messages")).toBeTruthy();
    });

    it("renders favourite messages correctly for a single event", () => {
        const props = {
            favouriteMessageEvents: [bobsFavouriteMessageEvent],
            handleSearchQuery: jest.fn(),
            searchQuery: "",
            cli,
        };
        const panel = render(<FavouriteMessagesPanel {...props} />);

        panel.getByText("i am bob");
    });

    it("renders favourite messages correctly for multiple single event", () => {
        const props = {
            favouriteMessageEvents: [alicesFavouriteMessageEvent, bobsFavouriteMessageEvent],
            handleSearchQuery: jest.fn(),
            searchQuery: "",
            cli,
        };

        const panel = render(<FavouriteMessagesPanel {...props} />);

        panel.getByText("i am alice");
        panel.getByText("i am bob");

        expect(panel.getAllByRole("link")[1].getAttribute("href")).toContain("$alices_message");
        expect(panel.getAllByRole("link")[3].getAttribute("href")).toContain("$bobs_message");

        expect(panel.asFragment()).toMatchSnapshot();
    });
});
