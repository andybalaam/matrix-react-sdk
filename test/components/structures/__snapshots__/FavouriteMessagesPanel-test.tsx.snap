// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouriteMessagesPanel renders favourite messages correctly for multiple single event 1`] = `
<DocumentFragment>
  <div
    class="mx_FavMessagesHeader"
  >
    <div
      class="mx_FavMessagesHeader_Wrapper"
    >
      <div
        class="mx_FavMessagesHeader_Wrapper_left"
      >
        <span
          class="mx_BaseAvatar"
          role="presentation"
        >
          <span
            aria-hidden="true"
            class="mx_BaseAvatar_initial"
            style="font-size: 16.900000000000002px; width: 26px; line-height: 26px;"
          >
            F
          </span>
          <img
            alt=""
            aria-hidden="true"
            class="mx_BaseAvatar_image"
            data-testid="avatar-img"
            src="data:image/png;base64,00"
            style="width: 26px; height: 26px;"
          />
        </span>
        <span>
          Favourite Messages
        </span>
      </div>
      <div
        class="mx_FavMessagesHeader_Wrapper_right"
      >
        <div
          aria-label="Search"
          class="mx_AccessibleButton mx_RoomHeader_button mx_RoomHeader_searchButton"
          role="button"
          tabindex="0"
        />
        <div
          aria-label="Clear"
          class="mx_AccessibleButton mx_RoomHeader_button mx_FavMessagesHeader_clearAllButton"
          role="button"
          tabindex="0"
        />
      </div>
    </div>
  </div>
  <div
    class="mx_AutoHideScrollbar mx_ScrollPanel mx_RoomView_searchResultsPanel mx_FavouriteMessages_scrollPanel"
    tabindex="-1"
  >
    <div
      class="mx_RoomView_messageListWrapper"
    >
      <ol
        aria-live="polite"
        class="mx_RoomView_MessageList"
        style="height: 400px;"
      >
        <li>
          <h2>
            Room: My room
          </h2>
        </li>
        <li
          data-scroll-tokens="$alices_message"
        >
          <ol>
            <div
              aria-label="Thu, Jan 1 1970"
              class="mx_DateSeparator"
              role="separator"
              tabindex="-1"
            >
              <hr
                role="none"
              />
              <h2
                aria-hidden="true"
              >
                Thu, Jan 1 1970
              </h2>
              <hr
                role="none"
              />
            </div>
            <li
              aria-atomic="true"
              aria-live="off"
              class="mx_EventTile mx_EventTile_lastInSection"
              data-event-id="$alices_message"
              data-has-reply="false"
              data-layout="false"
              data-scroll-tokens="$alices_message"
              data-self="false"
              tabindex="-1"
            >
              <div
                class="mx_DisambiguatedProfile"
              >
                <span
                  class="mx_Username_color6 mx_DisambiguatedProfile_displayName"
                  dir="auto"
                >
                  @alice:server.org
                </span>
              </div>
              <div
                class="mx_EventTile_line"
              >
                <a
                  aria-label="00:02"
                  href="#"
                />
                <div
                  class="mx_MTextBody mx_EventTile_content"
                >
                  <a
                    href="#/room/!room:server.org/$alices_message"
                  >
                    <span
                      class="mx_EventTile_body"
                      dir="auto"
                    >
                      i am alice
                    </span>
                  </a>
                </div>
                <div
                  aria-label="Message Actions"
                  aria-live="off"
                  class="mx_MessageActionBar"
                  role="toolbar"
                >
                  <div
                    aria-label="Favourite"
                    class="mx_AccessibleButton mx_MessageActionBar_iconButton"
                    data-testid="$alices_message"
                    role="button"
                    tabindex="0"
                  >
                    <div />
                  </div>
                  <div
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Options"
                    class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
                    role="button"
                    tabindex="-1"
                  >
                    <div />
                  </div>
                </div>
              </div>
            </li>
          </ol>
        </li>
        <li
          data-scroll-tokens="$bobs_message"
        >
          <ol>
            <div
              aria-label="Thu, Jan 1 1970"
              class="mx_DateSeparator"
              role="separator"
              tabindex="-1"
            >
              <hr
                role="none"
              />
              <h2
                aria-hidden="true"
              >
                Thu, Jan 1 1970
              </h2>
              <hr
                role="none"
              />
            </div>
            <li
              aria-atomic="true"
              aria-live="off"
              class="mx_EventTile mx_EventTile_lastInSection"
              data-event-id="$bobs_message"
              data-has-reply="false"
              data-layout="false"
              data-scroll-tokens="$bobs_message"
              data-self="false"
              tabindex="-1"
            >
              <div
                class="mx_DisambiguatedProfile"
              >
                <span
                  class="mx_Username_color3 mx_DisambiguatedProfile_displayName"
                  dir="auto"
                >
                  @bob:server.org
                </span>
              </div>
              <div
                class="mx_EventTile_line"
              >
                <a
                  aria-label="00:02"
                  href="#"
                />
                <div
                  class="mx_MTextBody mx_EventTile_content"
                >
                  <a
                    href="#/room/!room:server.org/$bobs_message"
                  >
                    <span
                      class="mx_EventTile_body"
                      dir="auto"
                    >
                      i am bob
                    </span>
                  </a>
                </div>
                <div
                  aria-label="Message Actions"
                  aria-live="off"
                  class="mx_MessageActionBar"
                  role="toolbar"
                >
                  <div
                    aria-label="Favourite"
                    class="mx_AccessibleButton mx_MessageActionBar_iconButton"
                    data-testid="$bobs_message"
                    role="button"
                    tabindex="0"
                  >
                    <div />
                  </div>
                  <div
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Options"
                    class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
                    role="button"
                    tabindex="-1"
                  >
                    <div />
                  </div>
                </div>
              </div>
            </li>
          </ol>
        </li>
      </ol>
    </div>
  </div>
</DocumentFragment>
`;
